aesd-char-driver/main.c

```diff
@@ -0,0 +1,349 @@
+char *incomplete_packet;
+size_t incomplete_packet_size = 0;
+
+MODULE_AUTHOR("Kevin Tom");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+
+
+
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *dev;
+    PDEBUG("open");
+    dev = container_of(inode->i_cdev,struct aesd_dev, cdev);
+    filp->private_data = dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0, no_byte_to_copy = 0;
+    size_t entry_offset_byte_rtn;
+    struct aesd_buffer_entry *aesd_entry_ret;
+    struct aesd_dev *driver_data = filp->private_data;
+
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+
+    retval = mutex_lock_interruptible(&driver_data->aesd_char_mut);
+    if(retval != 0)
+    {
+        PDEBUG("Could not aquire mutex");
+        return retval;
+    }
+    
+    aesd_entry_ret = aesd_circular_buffer_find_entry_offset_for_fpos(&driver_data->circular_buffer, *f_pos, &entry_offset_byte_rtn);
+    if(aesd_entry_ret == NULL)
+    {
+        *f_pos = 0;
+        PDEBUG("Nothing was read");
+        retval = 0;
+        goto out; 
+    }
+
+    //Calcualting number of bytes to copy
+    no_byte_to_copy = aesd_entry_ret->size - entry_offset_byte_rtn;
+    if(count < no_byte_to_copy)
+        no_byte_to_copy = count;
+    
+    *f_pos += no_byte_to_copy; //Updating file positin for next time
+
+    //Copying to user space
+    if( copy_to_user(buf, (aesd_entry_ret->buffptr + entry_offset_byte_rtn), no_byte_to_copy) )
+    {
+        retval = -EFAULT;
+        PDEBUG("copy_from_user() failed");
+        goto out;
+    }
+
+    //Assigning return value to number of bytes that was copied
+    retval = no_byte_to_copy;
+
+
+    out:
+    mutex_unlock(&driver_data->aesd_char_mut); //Unlocking mutex
+    return retval;
+}
+
+
+
+
+
+
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+
+    ssize_t retval = -ENOMEM;
+    char *user_krnl_buf = NULL, *packet = NULL, *c_buf_return;
+    size_t  size_of_current_packet, start_of_current_packet = 0;
+    int i=0;
+    struct aesd_dev *driver_data = filp->private_data;
+    
+    //Variable used for writing into buffer
+    struct aesd_buffer_entry buffer_entry;
+
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+
+
+    //filp->private_data.circular_buffer;
+    retval = mutex_lock_interruptible(&driver_data->aesd_char_mut);
+    if(retval != 0)
+    {
+        PDEBUG("Could not aquire mutex");
+        return retval;
+    }
+
+
+    //allocating space for user data in kernel space
+    user_krnl_buf = (char *)kmalloc(count,GFP_KERNEL);
+    if(user_krnl_buf == NULL)
+    {
+        PDEBUG("kmalloc() failed");
+        return retval;
```

Comment:
@kevintom98 what happens to your mutex if this step fails?

Also does this need to be inside the mutex locked region?